File: 139_dp.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
	bool wordBreak(string s, vector<string> &wordDict)
	{
		vector<bool> tab(s.length() + 1, false);
		tab[0] = true;

		for (int index = 0; index < tab.size() - 1; ++index)
		{
			if (!tab[index])
			{
				continue;
			}
			for (auto word : wordDict)
			{
				string chop(s.data() + index,
					    s.data() + min(index + word.length(), s.length()));
				if (word == chop)
				{
					tab[index + word.length()] = true;
				}
			}
		}

		for (auto i : tab)
		{
			cout << i << ' ';
		}
		cout << '\n';

		return tab.back();
	}
};

int main()
{
	string s = "catsandog";
	vector<string> wordDict = { "cats", "dog", "sand", "and", "cat" };

	Solution sol;

	sol.wordBreak(s, wordDict);

	return 0;
}

File: 139_dp.rs
struct Solution {}

impl Solution {
    pub fn word_break(s: String, word_dict: Vec<String>) -> bool {
        let mut tab = vec![false; s.len() + 1];
        tab[0] = true;

        for index in 0..tab.len() - 1 {
            if !tab[index] {
                continue;
            }

            for word in word_dict.iter() {
                let chop =
                    s[index..std::cmp::min(index + word.chars().count(), s.len())].to_string();
                if word.to_string() == chop {
                    tab[index + word.len()] = true;
                }
            }
        }

        println!("{:?}", tab);

        return tab.last().copied().unwrap();
    }
}

fn main() {
    let s = "catsandog".to_string();
    let word_dict = vec![
        "cats".to_string(),
        "dog".to_string(),
        "sand".to_string(),
        "and".to_string(),
        "cat".to_string(),
    ];

    let s = "leetcode".to_string();
    let word_dict = vec!["leet".to_string(), "code".to_string()];
    Solution::word_break(s, word_dict);
}
